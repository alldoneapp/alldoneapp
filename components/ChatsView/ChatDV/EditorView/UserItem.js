import React from 'react'
import { Image, StyleSheet, Text, TouchableOpacity, View } from 'react-native'
import HelperFunctions from '../../../../utils/HelperFunctions'
import Icon from '../../../Icon'
import global, { colors } from '../../../styles/global'

const UserItem = ({ user, selectUser, selected = false }) => {
    return (
        <View style={{ paddingHorizontal: 8 }}>
            <TouchableOpacity
                accessible={false}
                style={[localStyles.innerContainer, selected && { backgroundColor: 'rgba(139, 149, 167, 0.22)' }]}
                onPress={() => selectUser(user)}
            >
                <Image
                    source={{ uri: user.photoURL }}
                    style={[localStyles.userImage, selected && localStyles.circle]}
                />
                <Text style={[localStyles.members, selected && { color: colors.Primary100 }]}>
                    {HelperFunctions.getFNameFLastN(user.displayName)}
                </Text>
                {selected && <Icon name={'check'} size={24} color="#fff" style={{ marginLeft: 'auto' }} />}
            </TouchableOpacity>
        </View>
    )
}

const localStyles = StyleSheet.create({
    innerContainer: {
        flex: 1,
        flexDirection: 'row',
        alignItems: 'center',
        borderRadius: 4,
        paddingVertical: 8,
        paddingHorizontal: 8,
    },
    userImage: {
        height: 32,
        width: 32,
        borderRadius: 100,
        marginRight: 8,
    },
    circle: {
        borderWidth: 2,
        borderColor: colors.Primary100,
    },
    members: {
        ...global.subtitle1,
        color: 'white',
    },
})

export default UserItem
