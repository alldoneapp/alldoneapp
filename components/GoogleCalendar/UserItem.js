import React, { useState } from 'react'
import { Image, StyleSheet, Text, TouchableOpacity, View } from 'react-native'
import HelperFunctions from '../../utils/HelperFunctions'
import Icon from '../Icon'
import global, { colors } from '../styles/global'

const UserItem = ({ user, selectUser, all }) => {
    const [marked, setMarked] = useState(false)
    return (
        <View style={{ paddingHorizontal: 8 }}>
            <TouchableOpacity
                accessible={false}
                style={[localStyles.innerContainer, !all && marked && { backgroundColor: 'rgba(139, 149, 167, 0.22)' }]}
                onPress={() => {
                    setMarked(!marked)
                    selectUser(user)
                }}
            >
                <Image
                    source={{ uri: user.photoURL }}
                    style={[localStyles.userImage, !all && marked && localStyles.circle]}
                />
                <Text style={[localStyles.members, !all && marked && { color: colors.Primary100 }]}>
                    {HelperFunctions.getFNameFLastN(user.displayName)}
                </Text>
                {!all ? marked && <Icon name={'check'} size={24} color="#fff" style={{ marginLeft: 'auto' }} /> : null}
            </TouchableOpacity>
        </View>
    )
}

const localStyles = StyleSheet.create({
    innerContainer: {
        flex: 1,
        flexDirection: 'row',
        alignItems: 'center',
        borderRadius: 4,
        paddingVertical: 8,
        paddingHorizontal: 8,
    },
    userImage: {
        height: 32,
        width: 32,
        borderRadius: 100,
        marginRight: 8,
    },
    circle: {
        borderWidth: 2,
        borderColor: colors.Primary100,
    },
    members: {
        ...global.subtitle1,
        color: 'white',
    },
})

export default UserItem
