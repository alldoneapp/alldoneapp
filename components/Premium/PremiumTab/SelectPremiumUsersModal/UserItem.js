import React from 'react'
import { StyleSheet, Text, TouchableOpacity, View } from 'react-native'

import styles, { colors } from '../../../styles/global'
import Icon from '../../../Icon'
import SelectedAvatar from '../../../UIComponents/FloatModals/GoalAssigneesModal/SelectedAvatar'
import Avatar from '../../../Avatar'

export default function UserItem({ user, toggleSelection, isSelected, paidByOtherUser }) {
    const { displayName, photoURL, uid } = user

    const selectUser = () => {
        toggleSelection(isSelected, uid)
    }

    return (
        <View style={[localStyles.container, paidByOtherUser && { opacity: 0.5 }]}>
            <TouchableOpacity style={localStyles.userDataContainer} onPress={selectUser} disabled={paidByOtherUser}>
                <View style={localStyles.userData}>
                    {isSelected ? (
                        <SelectedAvatar photoURL={photoURL} />
                    ) : (
                        <Avatar
                            avatarId={uid}
                            reviewerPhotoURL={photoURL}
                            size={32}
                            borderSize={0}
                            externalStyle={{ marginRight: 8 }}
                        />
                    )}
                    <Text style={localStyles.text}>{displayName}</Text>
                </View>
                {(isSelected || paidByOtherUser) && <Icon name="crown" color={colors.Text03} size={24} />}
            </TouchableOpacity>
        </View>
    )
}

const localStyles = StyleSheet.create({
    container: {
        height: 48,
        flexDirection: 'row',
        alignItems: 'center',
        marginHorizontal: 8,
    },
    userDataContainer: {
        flexDirection: 'row',
        alignItems: 'center',
        flex: 1,
        justifyContent: 'space-between',
    },
    userData: {
        flexDirection: 'row',
        alignItems: 'center',
    },
    text: {
        ...styles.subtitle1,
        color: '#ffffff',
    },
})
