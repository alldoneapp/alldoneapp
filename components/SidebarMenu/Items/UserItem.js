import React, { useEffect } from 'react'
import { StyleSheet, TouchableOpacity, View } from 'react-native'
import { useDispatch, useSelector } from 'react-redux'

import store from '../../../redux/store'
import {
    hideWebSideBar,
    setSelectedSidebarTab,
    setSelectedTypeOfProject,
    setGoalsActiveTab,
    setTaskViewToggleIndex,
    setTaskViewToggleSection,
    storeCurrentShortcutUser,
    storeCurrentUser,
    setSelectedNavItem,
} from '../../../redux/actions'
import { GOALS_OPEN_TAB_INDEX } from '../../GoalsView/GoalsHelper'
import NavigationService from '../../../utils/NavigationService'
import {
    DV_TAB_ROOT_GOALS,
    DV_TAB_ROOT_TASKS,
    DV_TAB_USER_PROFILE,
    ROOT_ROUTES,
} from '../../../utils/TabNavigationConstants'
import { exitsOpenModals } from '../../ModalsManager/modalsManager'
import ContactsHelper from '../../ContactsView/Utils/ContactsHelper'
import { getUserItemTheme } from '../Themes'
import useCollapsibleSidebar from '../Collapsible/UseCollapsibleSidebar'
import useOnHover from '../../../hooks/UseOnHover'
import UserData from './Common/UserData'
import Indicator from './Common/Indicator'
import Amount from './Common/Amount'
import ProjectHelper from '../../SettingsView/ProjectsSettings/ProjectHelper'

export default function UserItem({
    user,
    projectType,
    projectId,
    projectColor,
    projectIndex,
    isShared,
    shortcut,
    navItem,
    showIndicator,
}) {
    const dispatch = useDispatch()
    const themeName = useSelector(state => state.loggedUser.themeName)
    const currentUserId = useSelector(state => state.currentUser.uid)
    const showShortcuts = useSelector(state => state.showShortcuts)
    const shownFloatPopup = useSelector(state => state.shownFloatPopup)
    const shortcutCurrentUserUid = useSelector(state => state.shortcutCurrentUserUid)
    const { expanded } = useCollapsibleSidebar()

    const theme = getUserItemTheme(themeName)

    const showShortcut = shortcut != null && showShortcuts && !shownFloatPopup && !exitsOpenModals()
    const highlight = currentUserId === user.uid
    const showAmount = navItem === DV_TAB_ROOT_TASKS
    const { hover, onHover, offHover } = useOnHover(highlight, highlight)

    const hideSideBar = () => {
        if (store.getState().smallScreenNavigation) dispatch(hideWebSideBar())
    }

    const getLastVisitedBoardProerty = () => {
        if (navItem === DV_TAB_ROOT_GOALS) return 'lastVisitBoardInGoals'
        if (navItem === DV_TAB_ROOT_TASKS) return 'lastVisitBoard'
    }

    const onPress = e => {
        e?.preventDefault()
        const { route, selectedNavItem } = store.getState()

        if (currentUserId === user.uid && (route === DV_TAB_ROOT_TASKS || route === DV_TAB_ROOT_GOALS)) {
            const project = ProjectHelper.getProjectById(projectId)
            NavigationService.navigate('UserDetailedView', {
                contact: user,
                project,
            })
            dispatch(setSelectedNavItem(DV_TAB_USER_PROFILE))
            return
        }

        if (selectedNavItem !== navItem) {
            dispatch(setSelectedSidebarTab(navItem))
        }

        if (!ROOT_ROUTES.includes(route)) NavigationService.navigate('Root')

        ContactsHelper.setUserLastVisitedBoardDate(projectId, user, getLastVisitedBoardProerty())

        let dispatches = [
            setSelectedSidebarTab(navItem), // Comment this to NOT go to Tasks when change user in sidebar
            storeCurrentUser(user),
            setSelectedTypeOfProject(projectType),
            storeCurrentShortcutUser(null),
        ]

        if (navItem === DV_TAB_ROOT_TASKS) {
            dispatches.push(setTaskViewToggleIndex(0))
            dispatches.push(setTaskViewToggleSection('Open'))
        } else if (navItem === DV_TAB_ROOT_GOALS) {
            dispatches.push(setGoalsActiveTab(GOALS_OPEN_TAB_INDEX))
        }

        dispatch(dispatches)
        hideSideBar()
    }

    useEffect(() => {
        if (shortcutCurrentUserUid === user.uid) onPress()
    }, [shortcutCurrentUserUid])

    return (
        <TouchableOpacity
            onPress={onPress}
            disabled={isShared}
            accessibilityLabel={'sidebar-user-item'}
            nativeID={`sidebar-user@${user.uid}`}
        >
            {/*{showShortcut && <ItemShortcut shortcut={shortcut} />}*/}
            <View
                style={[
                    localStyles.container,
                    highlight ? theme.containerActive(projectColor) : theme.container(projectColor),
                    !expanded && localStyles.containerCollapsed,
                    !highlight && hover && theme.containerActive(projectColor),
                ]}
                onMouseEnter={onHover}
                onMouseLeave={offHover}
            >
                {showIndicator && !highlight && <Indicator projectColor={projectColor} />}
                <UserData user={user} projectId={projectId} projectIndex={projectIndex} />
                {showAmount && !isShared && (expanded || highlight) && (
                    <Amount userId={user.uid} projectColor={projectColor} projectId={projectId} />
                )}
            </View>
        </TouchableOpacity>
    )
}

const localStyles = StyleSheet.create({
    container: {
        alignItems: 'center',
        justifyContent: 'space-between',
        flexDirection: 'row',
        height: 48,
        paddingLeft: 26,
    },
    containerCollapsed: {
        paddingLeft: 18,
    },
})
