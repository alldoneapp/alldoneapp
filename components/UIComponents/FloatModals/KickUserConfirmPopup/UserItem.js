import React from 'react'
import { Image, StyleSheet, Text, TouchableOpacity, View } from 'react-native'

import styles, { colors } from '../../../styles/global'
import Icon from '../../../Icon'

export default function UserItem({ user, selectedUserId, setSelectedUserId, inProgress }) {
    const { uid, photoURL, displayName } = user
    return (
        <TouchableOpacity onPress={() => setSelectedUserId(uid)} disabled={inProgress}>
            <View style={localStyles.userItem}>
                <View style={{ flexDirection: 'row', alignItems: 'center' }}>
                    <Image source={{ uri: photoURL }} style={localStyles.userImage} />
                    <Text style={[styles.subtitle1, { color: '#ffffff', marginLeft: 8 }]}>{displayName}</Text>
                </View>
                {selectedUserId === uid && <Icon name="check" size={24} color="white" />}
            </View>
        </TouchableOpacity>
    )
}

const localStyles = StyleSheet.create({
    userItem: {
        height: 48,
        paddingVertical: 12,
        flexDirection: 'row',
        justifyContent: 'space-between',
        alignItems: 'center',
    },
    userImage: {
        backgroundColor: colors.Text03,
        height: 32,
        width: 32,
        borderRadius: 100,
    },
})
