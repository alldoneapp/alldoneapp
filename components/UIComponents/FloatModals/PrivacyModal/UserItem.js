import React, { useState, useEffect } from 'react'
import { Image, StyleSheet, Text, TouchableOpacity, View } from 'react-native'

import styles, { colors, hexColorToRGBa } from '../../../styles/global'
import Icon from '../../../Icon'
import SVGGenericUser from '../../../../assets/svg/SVGGenericUser'
import { DEFAULT_WORKSTREAM_ID, getWorkstreamById, WORKSTREAM_ID_PREFIX } from '../../../Workstreams/WorkstreamHelper'
import { translate } from '../../../../i18n/TranslationService'
import TasksHelper from '../../../TaskListView/Utils/TasksHelper'

export default function UserItem({
    isOwner,
    ownerText,
    selected,
    hovered,
    onPress,
    itemsRefs,
    itemRefIndex,
    userId,
    projectId,
}) {
    const [user, setUser] = useState(null)

    useEffect(() => {
        const isWorkstream = userId.startsWith(WORKSTREAM_ID_PREFIX)
        const user = isWorkstream ? getWorkstreamById(projectId, userId) : TasksHelper.getPeopleById(userId, projectId)
        setUser(user)
    }, [userId])

    if (!user) return null

    const { uid, photoURL, displayName } = user

    return (
        <View ref={ref => (itemsRefs[itemRefIndex] = ref)}>
            <TouchableOpacity onPress={e => onPress(e, uid)} accessible={false}>
                <View style={[localStyles.elementItem, hovered && localStyles.itemSelected]}>
                    <View style={{ flex: 1, flexDirection: 'row', alignItems: 'center' }}>
                        {uid.startsWith(WORKSTREAM_ID_PREFIX) ? (
                            <View style={[localStyles.userImage, hovered && localStyles.userImageSelected]}>
                                <Icon size={24} name="workstream" color={colors.Text03} />
                            </View>
                        ) : photoURL ? (
                            <Image
                                source={{ uri: photoURL }}
                                style={[localStyles.userImage, hovered && localStyles.userImageSelected]}
                            />
                        ) : (
                            <View style={[localStyles.userImage, hovered && localStyles.userImageSelected]}>
                                <SVGGenericUser
                                    width={hovered ? 28 : 32}
                                    height={hovered ? 28 : 32}
                                    svgid={`av-priv-${uid}`}
                                />
                            </View>
                        )}
                        <View style={localStyles.itemTexts}>
                            <Text
                                style={[
                                    styles.subtitle1,
                                    localStyles.itemName,
                                    selected && localStyles.itemNameSelected,
                                ]}
                                numberOfLines={1}
                            >
                                {displayName !== ''
                                    ? uid === DEFAULT_WORKSTREAM_ID
                                        ? translate(displayName)
                                        : displayName
                                    : translate('Unknown user')}
                            </Text>
                            <Text
                                style={[
                                    styles.caption1,
                                    localStyles.itemSubName,
                                    (selected || isOwner) && localStyles.itemSubNameSelected,
                                ]}
                                numberOfLines={1}
                            >
                                {ownerText}
                            </Text>
                        </View>
                    </View>
                    {selected || isOwner ? (
                        <Icon name={'check'} size={24} color={isOwner ? colors.Text02 : 'white'} />
                    ) : (
                        <View style={{ width: 24, height: 24 }} />
                    )}
                </View>
            </TouchableOpacity>
        </View>
    )
}

const localStyles = StyleSheet.create({
    elementItem: {
        height: 48,
        paddingVertical: 12,
        flexDirection: 'row',
        justifyContent: 'space-between',
        alignItems: 'center',
        paddingLeft: 8,
        paddingRight: 8,
    },
    itemSelected: {
        backgroundColor: hexColorToRGBa(colors.Text03, 0.16),
        borderRadius: 4,
    },
    userImage: {
        flexDirection: 'row',
        justifyContent: 'center',
        alignItems: 'center',
        backgroundColor: 'transparent',
        height: 32,
        width: 32,
        borderRadius: 100,
        overflow: 'hidden',
    },
    userImageSelected: {
        borderWidth: 2,
        borderColor: colors.Primary100,
    },
    itemTexts: {
        flexDirection: 'row',
        alignItems: 'baseline',
    },
    itemName: {
        color: '#ffffff',
        marginLeft: 8,
    },
    itemSubName: {
        color: colors.Text04,
        marginLeft: 8,
    },
    itemSubNameSelected: {
        color: colors.Primary300,
    },
})
