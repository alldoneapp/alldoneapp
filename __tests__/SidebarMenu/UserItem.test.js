import React from 'react'
import UserItem from '../../components/SidebarMenu/UserItem'
import store from '../../redux/store'
import { storeLoggedUserProjects, switchProject, setNavigationRoute } from '../../redux/actions'

import renderer from 'react-test-renderer'
jest.mock('../../utils/NavigationService')

const navigationMock = {
    closeDrawer: () => {},
}

describe('UserItem component', () => {
    const user = { uid: '1', photoURL: 'https://dummy.com/dummy.png', displayName: 'John' }

    describe('UserItem snapshot test', () => {
        it('should render correctly', () => {
            const user = {
                uid: '1',
                displayName: 'Jhon Doe',
                photoURL: 'https://dummy.com/dummy.png',
            }
            const userItem = new UserItem({ user: user })

            const tree = renderer
                .create(<userItem user={userItem} tasksAmount={10} navigation={navigationMock} />)
                .toJSON()
            expect(tree).toMatchSnapshot()
        })
    })

    describe('test component methods call', () => {
        it('after call componentWillUnmount should render correctly', () => {
            const tree = renderer.create(<UserItem user={user} navigation={navigationMock} />)
            const instance = tree.getInstance()
            instance.componentWillUnmount()
        })

        it('after call updateState should render correctly', () => {
            const tree = renderer.create(<UserItem user={user} navigation={navigationMock} />)
            const instance = tree.getInstance()
            instance.updateState()
        })

        xit('should correctly handle navigation with browsing history', () => {
            const user = {
                uid: '1',
                displayName: 'Jhon Doe',
                photoURL: 'https://dummy.com/dummy.png',
            }
            const tree = renderer.create(<UserItem user={user} navigation={navigationMock} />)
            const instance = tree.getInstance()
            window.location = { pathname: '/down/the/rabbithole/with/alice', origin: 'https://wonderland.com' }
            window.hist = []
            history = {
                pushState: (state, title, url) => {
                    window.hist.push({ state: state, title: title, url: url })
                },
            }
            store.dispatch(setNavigationRoute('The Palace'))
            store.dispatch(switchProject(0))
            store.dispatch(storeLoggedUserProjects([{ name: 'Alice in Wonderland', id: '0' }]))
            instance.onPress()
            expect(window.hist[0]).toEqual({
                state: { projectId: '0', userId: user.uid },
                title: '',
                url: 'https://wonderland.com/down/1/rabbithole/with/alice',
            })
        })
    })
})
