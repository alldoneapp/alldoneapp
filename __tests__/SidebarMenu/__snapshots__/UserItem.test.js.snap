// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserItem component UserItem snapshot test should render correctly 1`] = `
<userItem
  navigation={
    Object {
      "closeDrawer": [Function],
    }
  }
  tasksAmount={10}
  user={
    UserItem {
      "_isMounted": false,
      "context": undefined,
      "firstFetch": true,
      "onPress": [Function],
      "onTaskEvent": [Function],
      "props": Object {
        "user": Object {
          "displayName": "Jhon Doe",
          "photoURL": "https://dummy.com/dummy.png",
          "uid": "1",
        },
      },
      "refs": Object {},
      "startWatchUserTasks": [Function],
      "state": Object {
        "currentUserId": undefined,
        "hiddenSideMenuUser": true,
        "loggedUserId": undefined,
        "searchText": "",
        "unsubscribe": [Function],
      },
      "updateState": [Function],
      "updater": Object {
        "enqueueForceUpdate": [Function],
        "enqueueReplaceState": [Function],
        "enqueueSetState": [Function],
        "isMounted": [Function],
      },
    }
  }
/>
`;
